File: src/main/java/me/anonymoussoftware/vacancymanager/model/Snippet.java
package me.anonymoussoftware.vacancymanager.model;

import org.json.JSONObject;

import lombok.AllArgsConstructor;
import lombok.Data;

@Data
@AllArgsConstructor
public class Snippet {

	private final String requirement;
	private final String responsibility;
	
	public static Snippet fromJson(JSONObject json) {
		if (json == null) {
			return new Snippet("", "");
		}
		return new Snippet(replaceUnknownHtmlTags(json.optString("requirement")), replaceUnknownHtmlTags(json.optString("responsibility")));
	}
	
	private static String replaceUnknownHtmlTags(String text) {
		return replaceUnknownTag(text, "highlighttext", "u");// text.replace("<highlighttext>", "<u>").replace("</highlighttext>", "</u>");
	}
	
	private static String replaceUnknownTag(String text, String tagName, String newTagName) {
		return text.replace("<" + tagName + ">", "<" + newTagName + ">").replace("</" + tagName + ">", "</" + newTagName + ">");
	}
	
	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder();
		if (this.requirement != null && !this.requirement.trim().isEmpty()) {
			sb.append("<b>Requirement:</b>");
			sb.append("<br/>");
			sb.append(this.requirement);
			sb.append("<br/>");
			sb.append("<br/>");
		}
		if (this.responsibility != null && !this.responsibility.trim().isEmpty()) {
			sb.append("<b>Responsibility:</b>");
			sb.append("<br/>");
			sb.append(this.responsibility);
		}
		return sb.toString();
	}

	public JSONObject toJson() {
		JSONObject result = new JSONObject();
		if (this.requirement != null && !this.requirement.trim().isEmpty()) {
			result.put("requirement", this.requirement);			
		}
		if (this.responsibility != null && !this.responsibility.trim().isEmpty()) {
			result.put("responsibility", this.responsibility);			
		}
		return result;
	}
}

File: src/main/kotlin/me/anonymoussoftware/vacancymanager/model/Snippet.kt
package me.anonymoussoftware.vacancymanager.model

import org.json.JSONObject

class Snippet (val requirement: String,
               val responsibility: String) : Model {
    override fun toString(): String {
        val sb = StringBuilder()
        if (this.requirement.isNotBlank()) {
            sb.append("<b>Requirement:</b><br/>$requirement<br/><br/>")
        }
        if (this.responsibility.isNotBlank()) {
            sb.append("<b>Responsibility:</b><br/>$responsibility")
        }
        return sb.toString()
    }

    fun toJson(): JSONObject {
        val result = JSONObject()
        if (this.requirement.isNotBlank()) {
            result.put("requirement", this.requirement)
        }
        if (this.responsibility.isNotBlank()) {
            result.put("responsibility", this.responsibility)
        }
        return result
    }

    companion object : JsonDeserializer<Model> {
        override fun fromJson(json: JSONObject): Snippet =
            Snippet(
                replaceUnknownHtmlTags(json.optString("requirement")),
                replaceUnknownHtmlTags(json.optString("responsibility")))

        private fun replaceUnknownHtmlTags(text: String): String =
            replaceUnknownTag(text, "highlighttext", "u")

        private fun replaceUnknownTag(text: String, tagName: String, newTagName: String) : String =
            text.replace("<$tagName>", "<$newTagName>")
                .replace("</$tagName>", "</$newTagName>")
    }
}
